gh-85222: Document the side effect in multiprocessing (#136426)
python/cpython: Doc/library/multiprocessing.rst

The :mod:`multiprocessing` package mostly replicates the API of the
:mod:`threading` module.

.. _global-start-method:

Global start method
^^^^^^^^^^^^^^^^^^^

> Calling some functions or methods, or creating some objects, will implicitly
> set the global start method to the system's default if it is not already
> set. This can only be done once. Therefore, if you need to change the
> start method, you must do so before calling these functions or methods, or
> creating these objects.
> 

Reviewer: ```suggestion
Python supports several ways to create and initialize a process. The global start method sets the default mechanism for creating a process.

Several multiprocessing functions and methods, as well as creating some objects, will implicitly
set the global start method to the system's default, if the global start method is not already
set. The global start method can only be done once. If you need to change the
start method from the system default, you must proactively set the global start method
before calling functions or methods, or creating these objects.

```

Reviewer 2: @willingc 

Good thinking... this adds a little background into what the Global Start Method is so people aren't lost

Reviewer: Thanks for the PR. I would love to see us be a bit more explicit here. Let me know if you have any questions. @gpshead please check that I'm not misrepresenting anything.

The :mod:`multiprocessing` package mostly replicates the API of the
:mod:`threading` module.

.. _global-start-method:

Global start method
^^^^^^^^^^^^^^^^^^^

Python supports several ways to create and initialize a process. The global start method sets the default mechanism for creating a process.

Several multiprocessing functions and methods that may also instantiate

> certain objects will implicitly set the global start method to the system's default, 

Reviewer 2: ```suggestion
certain objects will implicitly set the global start method to the system's default,
```

> set. The global start method can only be done once. If you need to change the

Author: ```suggestion
set. The global start method can only be set once. If you need to change the
```

@sharktide Hi, how about this?

Reviewer 2: sure ! 

EDIT: my other suggestion includes  this

> set. The global start method can only be done once. If you need to change the

Reviewer 2: Also, I think this improves the flow around where you say ``as well as creating some objects`` Also includes the grammatical edit you suggested because the lines were next to each other

```suggestion
Several multiprocessing functions and methods that may also instantiate
certain objects will implicitly set the global start method to the system's default, 
if it hasn’t been set already. The global start method can only be set once. If you need to change the
```

   locks/semaphores.  When a process first puts an item on the queue a feeder
   thread is started which transfers objects from a buffer into the pipe.

   If the global start method has not been set, calling this function will
   have the side effect of setting the current global start method.

>    See the :func:`get_context` function.

Reviewer 3: I think repeating the same words here may be a bit 'noisy'. Could we just mention it once?

Author: My concern is that if a user only uses one of these functions and checks that function's documentation, and unfortunately that is not the function that contains this document, they may mess it up.

But I'm unsure about this.

Reviewer 2: I may be biased but I don't think the extra clarity hurts... I'm with @aisk

Reviewer 4: yeah, I asked for this to be mentioned everywhere for that reason as this is non-obvious behavior of multiprocessing.  This could probably be refined a bit, i'll point some docs focused reviewers at it.

Reviewer 2: <html>
<body>
<!--StartFragment-->
ԅ(≖‿≖ԅ)
--

<!--EndFragment-->
</body>
</html>

sorry i just had to :)

Reviewer 3: > My concern is that if a user only uses one of these functions and checks that function's documentation, and unfortunately that is not the function that contains this document, they may mess it up.
> 
> But I'm unsure about this.

IMO it may be better if we mentioned it once and create a reference link to the mention in every function. I'm not sure too (｡･ω･｡)

Reviewer 5: The repeated part could be a single sentence like *Ensures that the current global start method is set.*, with ”global start method” linking to a longer explanation in a dedicated section.

Reviewer 6: I'm not sure it needs to be mentioned in every method.  There's a section at the top about contexts that can say the global start method is set by any function that does real work (or some appropriate language).  I find we often have overarching concerns that apply throughout a module and don't repeat them.  Sometime people have to read widely in a module page in order to understand all the nuances.

Maybe I missed the discussion: why is this particular caveat important enough to sprinkle everywhere? What footgun are we preventing?

Reviewer 4: The callout about the surprising behavior is also (in addition to the issue for this PR) due to https://github.com/python/cpython/issues/109070 which only _just_ added the caveat about `.get_context()` having the setting side effect behavior via https://github.com/python/cpython/pull/136341/files.

I agree that the https://docs.python.org/3.15/library/multiprocessing.html#contexts-and-start-methods section up top should be more clear about this. It really only ever mentions `get_context()` as an alternative today and never highlights that the context is implicitly set at instantiation time by all sorts of APIs...  We do say `To select a start method you use the set_start_method() in the if __name__ == '__main__' clause of the main module.` but we never actually explain the restrictions leading to that advice, the _why_, there...  That would help.

After that these mentions about context setting could turn into a single sentence similar to what encoku suggested that ref's back to that section.

_(the other part of this PR is documenting which things accept a ctx= parameter at all)_

Author: Hi, thank you all for review this PR, I updated the document and added a new section to explain the implicitly set of start method, and changed other methods/types to link to this section. Please review it to see if this is suitable.

   It is a simplified :class:`Queue` type, very close to a locked :class:`Pipe`.

>    If the global start method has not been set, calling this function will

Reviewer 5: After there's final wording for the repeated part is reworded, check that it's in all the contexts.

(This is class documentation; it should be “creating a `SimpleQueue` object” rather than “calling this function”.)

   :mod:`multiprocessing` module.

   If *method* is ``None`` then the default context is returned. Note that if

>    the global start method has not been set, this will set it.

Reviewer 2: ```suggestion
   the global start method has not been set, this will set it to the system default
```

Suggestion from one of my earlier comments

   Return the name of start method used for starting processes.

>    If the start method is not set and *allow_none* is ``False``, the start

Reviewer 2: ```suggestion
   If the global start method is not set and *allow_none* is ``False``, the global start
```

IMO it's better to mention "global start method" everywhere instead of "start method" for consistency

   automatically protected by a lock, so it will not necessarily be
   "process-safe".

   *ctx* is a context object, or ``None`` (use the current context). If ``None``,
   calling this function will have the side effect of setting the current global
   start method if it has not been set already. See the :func:`get_context`
   function.

>    Note that *lock* and *ctx* are keyword-only argument.

Reviewer 7: They are parameters, not arguments.

   automatically protected by a lock, so it will not necessarily be
   "process-safe".

   *ctx* is a context object, or ``None`` (use the current context). If ``None``,
   calling this function will have the side effect of setting the current global
   start method if it has not been set already. See the :func:`get_context`
   function.

> 

Reviewer 7: Same here, parameters, not arguments.

   the worker processes.  Usually a pool is created using the
   function :func:`multiprocessing.Pool` or the :meth:`Pool` method
   of a context object.  In both cases *context* is set
   appropriately. If ``None``, calling this function will have the side effect

>    of setting the current global start method if it has not been set already.

Reviewer 2: ```suggestion
   of setting the current global start method to the 
system default if it has not been set already.
```

More small suggestions to improve reading flow

Up to you if you wanna take this one
